src/pikepdf/models/outlines.py

```diff
@@ -0,0 +1,247 @@
+        """
+        if self.obj is None:
+            self.obj = obj = pdf.make_indirect(Dictionary())
+        else:
+            obj = self.obj
+        obj.Title = self.title
+        if self.destination is not None:
+            if isinstance(self.destination, int):
+                self.destination = get_page_destination(pdf, self.destination, self.page_location,
+                                                        **self.page_location_kwargs)
+            obj.Dest = self.destination
+            if '/A' in obj:
+                del obj.A
+        elif self.action is not None:
+            obj.A = self.action
+            if '/Dest' in obj:
+                del obj.Dest
+        return obj
+
+
+class Outline:
+    """Maintains a intuitive interface for creating and editing PDF document outlines,
+    according to the PDF reference manual (ISO32000:2008) section 12.3.
+
+    Arguments:
+        pdf: PDF document object.
+    """
+    def __init__(self, pdf):
+        self._root = []
+        self._pdf = pdf
+        if pdf:
+            self.load()
+
+    def __str__(self):
+        return str(self.root)
+
+    def __repr__(self):
+        return '<{0.__class__.__name__}: {1} items>'.format(self, len(self.root))
+
+    def _save_level_outline(self, parent: Dictionary, outline_items: list):
+        count = 0
+        prev = None
+        first = None
+        for item in outline_items:
+            out_obj = item.to_dictionary_object(self._pdf)
+            out_obj.Parent = parent
+            count += 1
+            if prev is not None:
+                prev.Next = out_obj
+                out_obj.Prev = prev
+            else:
+                first = out_obj
+                if '/Prev' in out_obj:
+                    del out_obj.Prev
+            prev = out_obj
+            self._save_level_outline(out_obj, item.children)
+            if item.is_closed:
+                out_obj.Count = -out_obj.Count
+            else:
+                count += out_obj.Count
+        if count:
+            if '/Next' in prev:
+                del prev.Next
+            parent.First = first
+            parent.Last = prev
+        else:
+            if '/First' in parent:
+                del parent.First
+            if '/Last' in parent:
+                del parent.Last
+        parent.Count = count
+
+    def _load_level_outline(self, first_obj: Dictionary, outline_items: list):
+        current_obj = first_obj
+        while current_obj:
+            item = OutlineItem.from_dictionary_object(current_obj)
+            first_child = current_obj.get('/First')
+            if first_child is not None:
+                self._load_level_outline(first_child, item.children)
+                count = current_obj.get('/Count')
+                if count and count < 0:
+                    item.is_closed = True
+            outline_items.append(item)
+            current_obj = current_obj.get('/Next')
+
+    def save(self):
```

Comment:
I suggest making `save` and `load` private, and having an `__enter__`/`__exit__` so that a series of changes to Outlines are performed in a `with` block. That would mirror what is done with `PdfMetadata`.

Reply:
I have added a similar implementation with open_outline() now.